Cities dev

jjgomezcadenas opened the pull request:
This PR simplifies the structure of the cities, making then one step closer to the goal of being truly invisibles. The most important features are:

1) Most cities have the same run() method, that now goes to the base (only Zaira needs to re-write the run() method).

2) Cities reading RWF (Isidora and Irene) have the same file_loop() method, that now goes to their common base.

3) Cities reading PMAPS (Dorothea and Penthiselea) have the same file_loop() method that now goes to their common base (Dorothea and Penthiselea both inherit now from HitCity).

4) A counter class allows easy bookeping and common go() method. 

Overall, the concrete cities are much thiner and simpler, with considerable reduction of noise and sizeable repayment of TD).

jjgomezcadenas committed:
Relax testing conditions to avoid superious failures in travis.

jjgomezcadenas committed:
Fix tolerances in hst_functions_test

jjgomezcadenas committed:
Cosmetics (legal and authors in Diomira).

jjgomezcadenas committed:
Cosmetics (add credits)

jjgomezcadenas committed:
Add type Counter

A new type, Counter has been added to ic_types. Counter allows to
define, increment and set an arbitrary list of counters.

jjgomezcadenas committed:
New package types created and files moved to it.

Package types has been created. All ic_types (both python and cython)
are now moved to this package.

In addition all modules related with event_modules are now in package evm.

jjgomezcadenas committed:
Move nh5 to evm

nh5.py contains the persistent definition of everything in IC (e.g,
the pytables classes) and belongs to the evm package. It has been
moved there and the code correspondingly adapted.

jjgomezcadenas committed:
Added authors and legal info to IC doc

jjgomezcadenas committed:
time and length are compulsory in find_s1(s2).

Accordingly they should not take a default. Fixed now.

jjgomezcadenas committed:
Make print_empty_events =1 the default in irene conf

jjgomezcadenas committed:
Minor modifications in params.py

Change name CalibratedSum to CSum
Add S12Sum

jjgomezcadenas committed:
Counter type can take a global name.

This allows to define the name of the instance (e.g, the city) holding
the counters.

jjgomezcadenas committed:
Add function get_mc_tracks to tbl_functions

Function get_mc_tracks returns the MC Tracks table. Function hides the
pytables implementation.

jjgomezcadenas committed:
Refactor compute_csum_and_pmaps in peak_functions.

Function compute_csum_and_pmaps has been refactore to make full use of
new pmaps interface.

jjgomezcadenas committed:
Refactor base_cities to improve structures of cities.

An important refactoring of base_cities is made to improve the
structure of the cities.

1. base_cities now holds the method run() which is the same for most
cities. The method run performs standard operations on files and data
base and needs to be over-written only occassionaly (e.g, in Zaira,
whose structure is different from that of other cities).

2. base cities also provides interfaces to be implemented by concrete
cities. These are:

file_loop()

event_loop()

display_IO_info()

write_parameters(h5out):

get_writers(h5out):

The run method returns a dictionary of counters which can be
manipulated after the city has run.

The function get_mc_tracks() has been added to the base, to
return MC tracks for MC events

function pmt_transformations() now returns a S12Sum and a CSum named
tuples.

jjgomezcadenas committed:
refactor irene and dorothea

The cities of irene and dorothea (together with tests) have been
refactored to use the new base_cities interface. The cities no longer
need the method run, and implement instead the interface defined in
the base.

jjgomezcadenas committed:
Add new method to base_cities

method write_simulation_parameters_table has been added to
base_cities, masking the direct call to the free functions API

jjgomezcadenas committed:
Mostly cosmetics.

mostly cosmetics changes to irene.

jjgomezcadenas committed:
Fix an accidental bug in irene_test

File was ended in h5.h5 due to an spurious error. Now fixed.

jjgomezcadenas committed:
New interface propagated to diomira, isidora & zaira.

The new interface (using counters, general run method, etc.) has been
propapagated to diomira, isidora and zaira

jjgomezcadenas committed:
Change def. config. for Dorothea

Dorothea now inherits from HitCity, the new configuration file
reflects this.

jjgomezcadenas committed:
Method file_loop added to the base

Method file_loop is the same for cities that read RWF and produce
PMAPS (e.g, Irene, Isidora). The cities that read PMAPS and produce
hits also have the same file_loop (Penthiselea, Dorothea), thus
file_loop is added to the correspoding base cities, simplifying
further the concrete cities.

HitCollectionCity is now named HitCity, for simplicity.

jjgomezcadenas committed:
Common file_loop method for irene and isidora

Irene and Isidora now use the same file_loop method, which goes to the
base, further simplfying the concrete cities and revealing their
common structure

jjgomezcadenas committed:
Dorothea and Penthiselea share same file_loop method.

Dorothea and Penthiselea now use the same file_loop method, which goes to the
base, further simplfying the concrete cities and revealing their
common structure

jmbenlloch committed:
Add header to cities

jmbenlloch committed:
Rename hit_collection_city.conf to hit_city.conf

jmbenlloch committed:
Add MIT license

jacg commented on invisible_cities/core/ic_types_test.py:
```diff
@@ -140,3 +141,82 @@ def test_cxy(xyc, a, b):
     np.isclose (xyc.Phi, phi, rtol=1e-4)
     np.allclose(xyc.pos, pos, rtol=1e-3, atol=1e-03)
+
+def test_counters():
```
Please break this test up into separate orthogonal tests, wherever possible.

jacg commented on invisible_cities/cities/dorothea.py:
```diff
@@ -44,61 +43,75 @@ def __init__(self, **kwds):
-                break
-            else:
-                print("OK")
+    # def file_loop(self):
```
Don't forget to remove these commented-out lines before the final merge.

jacg commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -14,87 +22,131 @@
-    def _event_loop(self, event_numbers, timestamps,
-                    nevt_in, nevt_out, write_hits, S1_dict, S2_dict, S2Si_dict):
-        max_events_reached = False
+    # def run(self):
```
More commented-out code here to be removed before final commit.

jmbenlloch committed:
Remove blocks of old commented code

jjgomezcadenas committed:
Remove dead code in irene.

jjgomezcadenas committed:
Improve Diomira.

1) Remove dead code.
2) Move file_loop to the base

jjgomezcadenas committed:
Removed dead code from Isidora and Zaira.

jjgomezcadenas committed:
Rename module params.py (reco) to ic_containers (evm).

Module params.py had the wrong name and lived in the wrong place. The
module defined named tuples (e.g, containers), and has been renamed to
ic_containers and placed in evm.

jjgomezcadenas committed:
Recover configuration file hit_collection_city.conf

jjgomezcadenas committed:
Propagate change (params.py --> ic_containers).

jjgomezcadenas committed:
Add DataVectors and PmapVectors to ic_containers

jjgomezcadenas committed:
Use of DataVectors and PmapVectors in cities.

Traffic of arguments in function events_loop for the various cities
expressed more concisely  with the user of DataVectors and PmapVectors
containers.

jjgomezcadenas committed:
Split HitCity into KrCity and HitCity

A KrCity computes and writes KrEvents, e.g, point-like events. This
means that the baricenter integrates all the time slices for a given
SiPM. Instead, a HitCity computes and writes HitEvents, in which the
baricenter (or any other position-finding algo) is computed
slice-by-slice. A HitCity is derived from a KrCity since both of them
use the same file_loop() function (which starts from pmaps).

jjgomezcadenas committed:
Add traceback after printout if except

Without the exit, the command-line test was passing with pretty greens
when the city was not found (e.g, a captured ModuleNotFound, print
that pytest captures and digests without telling anything and happy
ending in spite of wrong situation). Adding traceback still gives
printout when running in normal mode but provides diagnostics of the problem

Commit:
rebin_s2si uses new Peak class

jjgomezcadenas committed:
Cosmetics

jjgomezcadenas committed:
Fixed wrong module names in cecilia and penthesilea

jjgomezcadenas committed:
Run 10 events by dafault in Penthesilea

The first few events are rejected by selection in the test sample in
Penthesilea, running 10 events guarantees that the city runs fully.

jjgomezcadenas committed:
work in progress

Commit:
write sipm_ids_and_charges_in_slice with test

this function allows takes an s2sid_peak and slice number and returns
the sipm ids of sipms in that s2si, with their charges at slice number

Commit:
cythonize sipm_ids_and_charges_in_slice

Commit:
fix tests for rebin_s2si

jjgomezcadenas committed:
Add pmaps_functions_c to base.

This is needed to use the corresponding cython function rather than
previous python version.

jjgomezcadenas committed:
Return only non-empty slices in sipm_ids_and_charges...

jjgomezcadenas committed:
Add writer to penthesilea.

jjgomezcadenas committed:
work in progress on cecilia.

jjgomezcadenas committed:
Modify functions sipm_ids_and_charges_in_slice.

Include in the list only slices with positive charge.

jjgomezcadenas committed:
Fix function compute_xy_position

One has to take into account the possibility that the slice results in
an empty SiPM list. This is handled via SipmEmptyList exception.

jjgomezcadenas committed:
Take into account that clusters for a given slice_no can be empty.

In this case, the slice is looped away.

jjgomezcadenas committed:
Remove Cecilia from cities.

Cecilia is removed from cities. The simulation of the trigger will be
done shortly in Diomira.

jjgomezcadenas committed:
Add NN (NoNumber) with value -999999

Although it may be argued that we reinvent the wheel, the use of a
numerical NN is quite convenient when writing/reading pytables to/from
disk, as a proxy for NaN.

jjgomezcadenas committed:
Cluster takes std (not rms) as argument.

Cluster takes the std (not the rms) as argument. A number of
properties (Xrm and such) are modified accordingly.

jjgomezcadenas committed:
Various tests involving cluster fixed.

The change of definition in Cluster (takes std, not rms) requires to
fix a number of tests.

jjgomezcadenas committed:
Redefined conf. files for kr_city and hit_city.

hit_city inherits now from kr_city. Config. files modified accordingly.

jjgomezcadenas committed:
Add functionality to KrCity.

Dorothea is now derived from KrCity, which centralizes the operations
of a city which computes and writes a KrEvent. In particular, in a
KrCity the position of the particle is computed integrating over all
temporal samples in SiPMs, but otherwise using the same
reco-algorithms (configurable) that can be used in HitCity (where
position is computed slice-by-slice).

jjgomezcadenas committed:
Penthesilea.conf modified to include hit_city.

jjgomezcadenas committed:
Dorothea simplified.

Like other cities before her, Dorothea has been further simplified,
transfering most of its functionality to the base. In particular,
KrEvent is no longer created by Dorothea (the proper place to create a
KrEvent is in KrCity). Tests have been modified accordingly.

jjgomezcadenas committed:
Move critical functionality of Penthesilea to the base.

Penthesilea follows the same fate that other cities before, and gets
leaner and thinner (more invisible), transfering functionality to the
base. Following the existing model, it is the HitCity (and not a
derived slave) the one that creates the HitEvent.

jjgomezcadenas committed:
Eliminate "Persistent" classes.

The use of "Persistent" class is unnecessary and somewhat
confusing. Following the model of pmap classes, all other "persistent"
classes now include a method store that is used by the corresponding
writer for persistent storage. Thus, classes PersistentHitCollection
and PersistentKrEvent have been sent to Valhaya.

jjgomezcadenas committed:
Propagate elimination of "Persistent".

Several tests use now HitCollection and KrEvent rather than
PersistentHitCollection etc. Notice that instantiation to persistent
classes is no longer done in Dorothea or Penthesilea, but in the base.

jjgomezcadenas committed:
Cosmetics and cleaning old wood.

Dead code removed, some cosmetics applied.

jjgomezcadenas committed:
Fix a bug on dorothea.conf

Dorothea must call kr_city.con (it was calling hit_city.conf)

jjgomezcadenas committed:
Corona algorithm raises exceptions when needed.

Some times Corona may find an empty list, in which case an exception
is raised.

jjgomezcadenas committed:
Change name: test_xy_algo to xy_algo_test

To follow the standard convention

jmbenlloch reviewed (approved):
Important refactoring of the cities and general improvements. @abotas has been using this code in some notebooks and everything seems to work well.
